compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1535,19 +1506,38 @@ class Typer extends Namer
     if (sym is ImplicitOrImplied) checkImplicitConversionDefOK(sym)
     val tpt1 = checkSimpleKinded(typedType(tpt))
 
-    var rhsCtx = ctx
-    if (sym.isConstructor && !sym.isPrimaryConstructor && tparams1.nonEmpty) {
-      // for secondary constructors we need a context that "knows"
-      // that their type parameters are aliases of the class type parameters.
-      // See pos/i941.scala
-      rhsCtx = ctx.fresh.setFreshGADTBounds
-      (tparams1, sym.owner.typeParams).zipped.foreach { (tdef, tparam) =>
-        val tr = tparam.typeRef
-        rhsCtx.gadt.addBound(tdef.symbol, tr, isUpper = false)
-        rhsCtx.gadt.addBound(tdef.symbol, tr, isUpper = true)
+    val rhsCtx: Context = {
+      var _result: FreshContext = null
+      def resultCtx(): FreshContext = {
```

Comment:
Can we factor this better? I am getting a bit frustrated how much complexity is increased by the GADT stuff. Right now it is a tradeoff: Slightly more expressive GADTs is bought by a significant drop in maintainability of the compiler as a whole (yes it is that bad!). We should not be forced to make that tradeoff.

Reply:
This block was more complex than the one before, because I tried to not unnecessarily instantiate  a fresh context. I reverted it to the old version and further simplified both branches to just a single call to a GadtConstraint method.